hyper_reqwest/src/main.rs

```diff
@@ -0,0 +1,134 @@
+    let local_client = LocalClient::new();
+    let remote_client = RemoteClient::new(&url);
+    let app = App::new(remote_client);
+    rwcst::run(local_client, app).await;
+}
+
+struct LocalClient {
+    client: reqwest::Client,
+}
+
+struct RemoteClient {
+    client: reqwest::Client,
+    remote: String,
+}
+
+struct App {
+    info: Arc<Mutex<rwcst::Info>>,
+    client: RemoteClient,
+}
+
+#[derive(Debug, derive_more::From)]
+enum Err {
+    Server(hyper::Error),
+    Client(reqwest::Error),
+    Parsing(rwcst::ParsingError),
+}
+type Result<T> = std::result::Result<T, Err>;
+
+#[async_trait::async_trait(?Send)]
+impl rwcst::LocalClientImpl for LocalClient {
+    type Err = Err;
+
+    fn new() -> Self {
+        LocalClient { client: reqwest::Client::new() }
+    }
+
+    async fn fetch_info(&mut self) -> Result<rwcst::Info> {
+        Ok(self.client.get("http://localhost:8001").send().await?.json().await?)
+    }
+}
+
+#[async_trait::async_trait(?Send)]
+impl rwcst::RemoteClientImpl for RemoteClient {
+    type Err = Err;
+
+    fn new(remote: &str) -> Self {
+        RemoteClient { client: reqwest::Client::new(), remote: remote.to_owned() }
+    }
+
+    async fn fetch_package(&mut self) -> Result<Option<(rwcst::Package, rwcst::Signature)>> {
+        let response = self.client.get(&self.remote).send().await?;
+
+        if let reqwest::StatusCode::OK = response.status() {
+            let sign = rwcst::Signature::from_base64_str(
+                &response.headers().get("Signature").unwrap().to_str().unwrap(),
+            );
+            let pkg = rwcst::Package::parse(&response.bytes().await?)?;
+            return Ok(Some((pkg, sign)));
+        }
+
+        Ok(None)
+    }
+}
+
+#[async_trait::async_trait(?Send)]
+impl rwcst::AppImpl for App {
+    type Err = Err;
+    type RemoteClient = RemoteClient;
+
+    fn new(client: RemoteClient) -> Self {
+        let info = Arc::default();
+        App { info, client }
+    }
+
+    fn serve(&mut self) -> Result<()> {
+        async fn handle(req: Request<Body>, state: Arc<Mutex<rwcst::Info>>) -> Response<Body> {
+            match (req.method(), req.uri().path()) {
+                (&Method::GET, "/") => {
+                    use std::ops::Deref;
+
+                    let state = state.lock().await;
+                    let body = serde_json::to_string(&state.deref()).unwrap();
+                    Response::new(Body::from(body))
+                }
+                _ => {
+                    Response::builder().status(StatusCode::NOT_FOUND).body(NOTFOUND.into()).unwrap()
+                }
+            }
+        }
+        let state = self.info.clone();
+        let make_svc = make_service_fn(move |_conn| {
+            let state = state.clone();
+            async move {
+                Ok::<_, Infallible>(service_fn(move |req: Request<Body>| {
+                    let state = state.clone();
+                    async move { Ok::<_, Infallible>(handle(req, state).await) }
```

Comment:
The handle function might fail if the route requested is no the expected one, and yet, here is set to a `Result<_, Infallible>`. I think the `handle` function should be reworked to return a `Result`.

Reply:
Done!